spec/consensus/proposer-based-timestamp/tla/TendermintPBT_002_draft.tla

```diff
@@ -394,92 +434,120 @@ ReceiveProposal(p) ==
-               \/ /\ proposalReceivedTime[r] /= NilTimestamp
-                  /\ UNCHANGED proposalReceivedTime
-         \/ /\ \/ localClock[p] - Precision >= t
-               \/ t >= localClock[p] + Precision + Delay
-            /\ UNCHANGED <<receivedTimelyProposal, proposalReceivedTime>>
-      /\ UNCHANGED <<round, step, decision, lockedValue, lockedRound,
-          validValue, validRound, evidence, msgsPropose, msgsPrevote, msgsPrecommit,
-          localClock, realTime, beginConsensus, endConsensus, lastBeginConsensus, proposalTime>>
+      /\ inspectedProposal[r,p] = NilTimestamp
+      /\ msg \notin receivedTimelyProposal[p]
+      /\ inspectedProposal' = [inspectedProposal EXCEPT ![r,p] = localClock[p]]
+      /\ LET 
+          isTimely == IsTimely(localClock[p], t)
+         IN
+          \/ /\ isTimely
+             /\ receivedTimelyProposal' = [receivedTimelyProposal EXCEPT ![p] = @ \union {msg}]
+             /\ LET
+                  isNilTimestamp == proposalReceivedTime[r] = NilTimestamp
+                IN
+                  \/ /\ isNilTimestamp
+                     /\ proposalReceivedTime' = [proposalReceivedTime EXCEPT ![r] = realTime]
+                  \/ /\ ~isNilTimestamp
+                     /\ UNCHANGED proposalReceivedTime
+          \/ /\ ~isTimely
+             /\ UNCHANGED <<receivedTimelyProposal, proposalReceivedTime>>
+      /\ UNCHANGED <<temporalVars, coreVars>>
+      /\ UNCHANGED
+        <<msgsPropose, msgsPrevote, msgsPrecommit, 
+        evidence(*, receivedTimelyProposal, inspectedProposal*)>>
+      /\ UNCHANGED 
+        <<beginRound, endConsensus, lastBeginRound,
+        proposalTime(*, proposalReceivedTime*)>>
       /\ action' = "ReceiveProposal"
       
 \* lines 22-27
 \* @type: (PROCESS) => Bool;
 UponProposalInPropose(p) ==
   \E v \in Values, t \in Timestamps:
+    LET 
+      r == round[p]
+    IN LET
+      \* @type: PROPOSAL;
+      prop == Proposal(v,t,r)
+    IN
     /\ step[p] = "PROPOSE" (* line 22 *)
     /\ LET 
         \* @type: PROPMESSAGE; 
         msg ==
         [
           type       |-> "PROPOSAL", 
-          src        |-> Proposer[round[p]],
-          round      |-> round[p], 
-          proposal   |-> Proposal(v, t), 
+          src        |-> Proposer[r],
+          round      |-> r, 
+          proposal   |-> prop, 
           validRound |-> NilRound
         ] 
        IN
-      /\ <<p, msg>> \in receivedTimelyProposal \* updated line 22
+      /\ msg \in receivedTimelyProposal[p] \* updated line 22
       /\ evidence' = {msg} \union evidence
     /\ LET mid == (* line 23 *)
-         IF IsValid(v) /\ (lockedRound[p] = NilRound \/ lockedValue[p] = v)
-         THEN Id(Proposal(v, t))
+         IF IsValid(prop) /\ (lockedRound[p] = NilRound \/ lockedValue[p] = v)
+         THEN Id(prop)
          ELSE NilProposal
        IN
-       BroadcastPrevote(p, round[p], mid) \* lines 24-26
+       BroadcastPrevote(p, r, mid) \* lines 24-26
```

Comment:
Not clear to me that we prevote nil if `msg` is not timely. Maybe move the `msg \in receivedTimelyProposal[p]` check in `LET mid...`??

Reply:
If `msg` is not timely, the entire action `UponProposalInPropose` is blocked.

Reply:
But we want to prevote nil, right?